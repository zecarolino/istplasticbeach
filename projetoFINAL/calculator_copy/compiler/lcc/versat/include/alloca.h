#ifndef _ALLOCA_H_
#define _ALLOCA_H_

#ifndef str
#define str(s) #s /* stringify */
#endif

/* function must have a frame (arguments or locals) */

/* 'siz' must be a global var and its value is replaced by the result */
#define alloca(siz) \
	asm("\tldi " str(siz) "\n\twrw RB\n\trdwb\n\trdwb\n\twrw RB\n" \
		"\trdw R12\n\tsub RB\n\twrw R12\n" \
		"\tldi " str(siz) "\n\twrw RB\n\trdw R12\n\twrwb\n")

/* 'pos' must be a global var and its value is set to sp=R12 position */
#define sp(pos) \
	asm("\tldi " str(pos) "\n\twrw RB\n\trdw R12\n\twrwb\n")

#endif /* _ALLOCA_H_ */
