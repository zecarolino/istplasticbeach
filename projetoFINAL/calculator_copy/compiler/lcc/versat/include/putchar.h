#ifndef _PUTCHAR_H_
#define _PUTCHAR_H_

void putchar(int ch) { asm("\trdw R11\n\taddi 2\n\twrw RB\n\trdwb\n\trdwb\n\twrw CPRT_BASE\n"); }

#define getchar #error "no getchar in versat"

#endif /* _PUTCHAR_H_ */
