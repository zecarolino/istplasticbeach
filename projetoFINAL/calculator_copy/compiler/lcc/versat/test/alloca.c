#include "putchar.h"
#include "alloca.h"

void printhex(int x) {
	int vec[] = { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F' };
	if (x) printhex(x >> 4);
	putchar(vec[x & 0xF]);
}


int pos;
void func(int ch) { /* must have a local */
	ch = ' ';
	sp(pos);
	printhex(pos);
	putchar(ch);
	pos = 12;
	alloca(pos);
	printhex(pos);
	putchar(ch);
	sp(pos);
	printhex(pos);
	ch = '\n';
	putchar(ch);
}

int main() { /* main can not take arguments */
	func(0);
	return 0;
}
