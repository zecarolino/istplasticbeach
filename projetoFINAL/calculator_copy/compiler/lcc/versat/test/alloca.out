VCD info: dumpfile xtop.vcd opened for output.
03FFB 03FEF 03FEF
0
Catched Trap at time  8471 clock cycles:
